sympy/core/numbers.py

```diff
@@ -425,6 +425,10 @@ def igcd_lehmer(a, b):
 def ilcm(*args):
     """Computes integer least common multiple.
 
+    It is retained for historical reasons,
+    but if there are no special needs,
+    please use ``sympy.external.gmpy.lcm``.
```

Comment:
I disagree with this comment being here. `ilcm` is exported to users. We should not be telling users that this is deprecated and that they should use an internal SymPy function instead. If anything, this function should just be modified to call gmpy.lcm. Why wasn't that done?

Reply:
And indeed I don't see why we need to change most code in SymPy to directly use the gmpy function instead of this one. Unless the code is already manipulating gmpy integers instead of SymPy integers for performance reasons, it is simpler to use this function, and avoids accidentally returning a gmpy mpz to the user instead of Integer.

Reply:
Like I said, the simplest way here is to just make this function call gmpy.lcm. Not only does this benefit users who use this function directly, but it makes it unnecessary for most SymPy functions to worry about using gmpy. The other benefit is that any function that could potentially return the result of lcm to the user will definitely return an Integer type if it is coming from this function. If they return the result of the backend `lcm`, it will be an `int` or `mpz` depending on the backend. Only very low level routines that don't operate on SymPy core types at all should use the gmpy functions directly, but that should be done explicitly.

Reply:
Discussion Point
Where do we absorb the difference between `int` and `mpz`? The function caller or the function called?

Ideal
The general user should not be aware of `int` and `mpz`, but the sympy developer should be. For example,
```python
def public_function(a):
    for _ in range(1000):
        a = call_gmpy_or_other_function(a)
    return a
```
In this case, it is unreasonable to expect `call_gmpy_or_other_function` to always return `int`. It is better to ensure consistency between the types at the beginning and end of `public_function`.

Reality
Since many existing codes do not assume `mpz`, the surest and easiest way is for the called function to absorb the difference between `int` and `mpz`, as you mentioned. On the other hand, there are many difficulties in trying to achieve the "ideal", and it is not a simple matter.

Conclusion
I think there is a realistic way to fix it immediately and an ideal way that takes a long time, both of which are options. It's hard to choose.

Reply:
I think that any code that uses the functions from sympy.external should be responsible for converting back to int if and when needed. There is a performance penalty in doing it when it is unnecessary that we will not want to pay for every possible use of things like `gmpy2.gcd`.

Ideally static type hints and a type checker should be used to keep track of whether functions are returning the expected types rather than doing anything that has a runtime performance impact.

Reply:
> I think that any code that uses the functions from sympy.external should be responsible for converting back to int if and when needed. 

Yes, and those functions will also be the ones to know that their performance matters enough to use the faster function. Everything else, whether it's user code or just a SymPy function that needs a simple gcd, will just use the Integer version. In the instances where you aren't calling it in a tight loop, the upside of returning an actual SymPy type outweighs the downside of the slight performance penalty of doing that conversion. 

And even without consideration of this discussion, we should still remove the comment from the docstring here, and make it so that ilcm/igcd just call the fast versions directly. @haru-44 can you make a PR to fix this?